submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,165 @@
+is enabled across all the clusters in a ClusterSet; That is all clusters that join a broker must use the same cable
+driver. The goal of this proposal is to extend the current model to support the selection of different cable types
+between connected clusters.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+ClusterSet. For example, a ClusterSet could contain three clusters, two that are on-premise and the other
+in a public cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+
+## Proposal
+
+This enhancement proposes a new policy to allow users to selectively deploy different cable types between clusters
+in a ClusterSet. The proposal assumes that connection between two clusters is symmetrical. If Cluster A is connected
+to Cluster B via a VXLAN cable type - then Cluster B is connected to Cluster A via a VXLAN cable type and not another
+cable type.
+
+This enhancement will propose a single API for cable driver policy support using custom resource definition (CRDs).
+
+### Cable driver policy API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    ClusterConnectionPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectionPoliciesSpec struct {
+ 
+    Policies []ClusterConnectionPolicy `json:"policies,omitempty"`
+}
+
+type ClusterConnectionPolicy struct {
+    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
+    // selector indicates wildcard, ie matches any cluster.
+    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`
+
+    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
+    // An empty selector indicates wildcard, ie matches any cluster.
+    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`
+
+    // Name of the cable driver implementation to use for this connection.
+    CableDriverName string `json:"CableDriverName"`
+
+    // CableOptions specifies a additional configuration parameters for the cable driver.
+    CableOptions map[string]string `json:"CableOptions,omitempty"`
+}
+```
+
+The `ClusterConnectionPolicies` CR defines the criteria for selecting which cable driver to use to connect cluster pairs (or subsets)
+and their associated (optional) configuration options. To resolve the connectivity policy for a remote cluster's Endpoint,
+Submariner would:
+
+- Search for a policy whose left and right label selectors match the corresponding Cluster resources of the local
```

Comment:
The suggestion here would be that the gw at startup time pulls in the connection policy CRs, and builds a local connection map[string]string , where the key is a cluster identifier or potentially even an endpoint identifier and the value is cable driver. On the first pass the default policy is applied to all values in the map. 

The next thing the GW should do is parse the rest of the policies - where it finds a policy whose left selectors match it, it establishes the list of clusters that match the right selectors and updates the connectivity map with what's specified in the policy.

When the time comes to actually install the cable - the connectivity map is consulted...

Reply:
rather than doing constant lookups of the CRs